README.md

# Chia log parser

Parses chia plot and harvest logs

## Install

	npm i chia-log-parser

## Examples

### Plot Parser

```javascript
const {plotParser} = require('chia-log-parser');

const parser = plotParser('/Users/home/.chia/mainnet/plotter/plot_log_xxx');
['phaseStart', 'phaseEnd', 'error', 'started', 'finished', 'done', 'progress'].forEach((event) => {
	parser.on(event, (...args) => console.log(`[${event}]`, ...args));
});

// if you only want to go by event
parser.start();

// get the full data
const plotData = await parser.parse();

// monitor log file in progress
parser.watch();
```

**Parsed output**

```javascript
	{
	  tmpDirs: [ '/tmp', '/tmp' ],
	  id: 'xxxxx',
	  plotSize: 32,
	  bufferSize: 3390,
	  buckets: 128,
	  threads: 2,
	  stripeSize: 65536,
	  phase1: {
	    startTime: 2021-05-08T09:38:18.000Z,
	    seconds: 9998.132,
	    cpuPercent: 145.15,
	    endTime: 2021-05-08T12:24:56.000Z
	  },
	  phase2: {
	    startTime: 2021-05-08T12:24:56.000Z,
	    seconds: 4188.688,
	    cpuPercent: 98.23,
	    endTime: 2021-05-08T13:34:45.000Z
	  },
	  phase3: {
	    startTime: 2021-05-08T13:34:45.000Z,
	    seconds: 7607.997,
	    cpuPercent: 98.21,
	    endTime: 2021-05-08T15:41:33.000Z
	  },
	  phase4: {
	    startTime: 2021-05-08T15:41:33.000Z,
	    seconds: 552.296,
	    cpuPercent: 99.46,
	    endTime: 2021-05-08T15:50:45.000Z
	  },
	  finalFileSize: 101.42,
	  totalTimeSeconds: 22347.115,
	  cpu: 119.24,
	  finishedTime: 2021-05-08T15:50:45.000Z,
	  copyTimeSeconds: 152.474,
	  copyCpu: 75.12,
	  copyFinishedTime: 2021-05-08T15:53:19.000Z
	}
```

### Harvest Parser

```javascript
const {harvestParser} = require('chia-log-parser');
const parser = harvestParser('/Users/foobar/.chia/mainnet/log/debug.log');

['load', 'warning', 'signagePoint'].forEach((event) => {
	parser.on(event, (...args) => console.log(`[${event}]`, ...args));
});

parser.watch().catch(console.log);
// [signagePoint] {
//   timestamp: 2021-05-25T16:36:16.305Z,
//   eligible: 0,
//   hash: '3ed0740787',
//   proofs: 0,
//   duration: 0.00487,
//   plots: 73
// }
// [warning] {
//   timestamp: 2021-05-21T09:11:48.229,
//   message: 'Directory: /Volumes/foobar/plots does not exist.'
// }
// [load] {
//   plots: 66,
//   size: 6.532919835822213,
//   sizeUnit: 'TiB',
//   time: 0.08295607566833496,
//   timestamp: new Date('2021-05-22T15:37:57.822')
// }

// parse from start to end
parser.parse().then((data) => {
	// data =
	// {
	//   load: [...],
	//   signagePoint: [...],
	//   warning: [...]
	// }
});
```


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

#### [2.0.1](https://github.com/martinj/chia-log-parser/compare/2.0.0...2.0.1)

- Fix issue with harvestParser.watch() failing on log file roll [`737adeb`](https://github.com/martinj/chia-log-parser/commit/737adeb77b075bf6985dcb83659715f5481e839a)

### [2.0.0](https://github.com/martinj/chia-log-parser/compare/1.0.0...2.0.0)

> 25 May 2021

- Export {plotParser, harvestParser} [`13cc76d`](https://github.com/martinj/chia-log-parser/commit/13cc76d1f0a05c7e2284db27b33311fcfd118d03)
- Rewrite and fix issues with plot parser not catching all lines on .watch() [`428398d`](https://github.com/martinj/chia-log-parser/commit/428398da5314bcb7227202caf7d1f09aa55770a7)
- Add harvest parser [`a1518e4`](https://github.com/martinj/chia-log-parser/commit/a1518e41c3b6a53196b83e3aca990988a557d60a)

### [1.0.0](https://github.com/martinj/chia-log-parser/compare/0.0.1...1.0.0)

> 20 May 2021

- Add unit tests and release scripts [`a21aac2`](https://github.com/martinj/chia-log-parser/commit/a21aac243d0fc6b50fdc5283d98fbda5130bfbbe)
- Add modified, created. Emit parseEnd with data and handle partial log files [`ba770c9`](https://github.com/martinj/chia-log-parser/commit/ba770c9ac475082e06e702ffc001642d9cb3a13c)

#### 0.0.1

> 19 May 2021

- Update ci config [`c2dfce7`](https://github.com/martinj/chia-log-parser/commit/c2dfce73474ca8dada877e2a921c4428f8f93832)
- Create node.js.yml [`dc58b1f`](https://github.com/martinj/chia-log-parser/commit/dc58b1f9f2c9bcb83b201c7d9072f90f440ea680)
- First commit [`03de1e7`](https://github.com/martinj/chia-log-parser/commit/03de1e70bd94ffe83d77527af780a977d72846ad)


test/plot.test.js
'use strict';

const expect = require('expect');
const plot = require('../src/plot');
const logFile = __dirname + '/fixtures/plot-log.txt';
const logFilePart = __dirname + '/fixtures/plot-log-unfinished.txt';

describe('Plot parser', () => {

	describe('stringToDate', () => {

		it('handles double and single digit days', () => {
			expect(plot.stringToDate('Fri May 21 13:38:03 2021')).toEqual(new Date('2021-05-21T11:38:03.000Z'));
			expect(plot.stringToDate('Sat May  8 11:38:18 2021')).toEqual(new Date('2021-05-08T09:38:18.000Z'));
		});

	});

	describe('parse()', () => {

		it('parses log file', async () => {
			const parser = plot(logFile);
			const res = await parser.parse();

			expect(res.modified).toBeInstanceOf(Date);
			expect(res.created).toBeInstanceOf(Date);
			delete res.created;
			delete res.modified;

			expect(res).toEqual({
				tmpDirs: ['/foobar', '/foobar'],
				id: '1553a1f4a9bb4d36a8a81e4334854c36d8abec1f192c362a28bd935fb035a03c',
				plotSize: 32,
				bufferSize: 3390,
				buckets: 128,
				threads: 2,
				stripeSize: 65536,
				phase1: {
					startTime: new Date('2021-05-08T09:38:18.000Z'),
					seconds: 9998.132,
					cpuPercent: 145.15,
					endTime: new Date('2021-05-08T12:24:56.000Z')
				},
				phase2: {
					startTime: new Date('2021-05-08T12:24:56.000Z'),
					seconds: 4188.688,
					cpuPercent: 98.23,
					endTime: new Date('2021-05-08T13:34:45.000Z')
				},
				phase3: {
					startTime: new Date('2021-05-08T13:34:45.000Z'),
					seconds: 7607.997,
					cpuPercent: 98.21,
					endTime: new Date('2021-05-08T15:41:33.000Z')
				},
				phase4: {
					startTime: new Date('2021-05-08T15:41:33.000Z'),
					seconds: 552.296,
					cpuPercent: 99.46,
					endTime: new Date('2021-05-08T15:50:45.000Z')
				},
				finalFileSize: 101.42,
				totalTimeSeconds: 22347.115,
				cpu: 119.24,
				finishedTime: new Date('2021-05-08T15:50:45.000Z'),
				copyTimeSeconds: 152.474,
				copyCpu: 75.12,
				copyFinishedTime: new Date('2021-05-08T15:53:19.000Z')
			});
		});

		it('parses non finished file', async () => {
			const parser = plot(logFilePart);
			const res = await parser.parse();

			expect(res.modified).toBeInstanceOf(Date);
			expect(res.created).toBeInstanceOf(Date);
			delete res.created;
			delete res.modified;
			expect(res).toEqual({
				tmpDirs: ['/foobar', '/foobar'],
				id: '1553a1f4a9bb4d36a8a81e4334854c36d8abec1f192c362a28bd935fb035a03c',
				plotSize: 32,
				bufferSize: 3390,
				buckets: 128,
				threads: 2,
				stripeSize: 65536,
				phase1: {startTime: new Date('2021-05-08T09:38:18.000Z')}
			});
		});

	});

	describe('start()', () => {

		it('emits events', (done) => {
			const parser = plot(logFile);
			const events = {
				phaseStart: 0,
				phaseEnd: 0,
				error: 0,
				started: 0,
				finished: 0,
				done: 0,
				progress: 0
			};

			Object.keys(events).forEach((event) => {
				parser.on(event, (...args) => {
					events[event]++;
					if (event === 'done') {
						expect(events).toEqual({
							phaseStart: 4,
							phaseEnd: 4,
							error: 0,
							started: 1,
							finished: 1,
							done: 1,
							progress: 19
						});
						done();
					}
				});
			});

			parser.start();
		});

		it('emits parseEnd for non finished log file', (done) => {
			const parser = plot(logFilePart);
			parser.on('parseEnd', (data) => {
				delete data.created;
				delete data.modified;

				expect(data).toEqual({
					tmpDirs: ['/foobar', '/foobar'],
					id: '1553a1f4a9bb4d36a8a81e4334854c36d8abec1f192c362a28bd935fb035a03c',
					plotSize: 32,
					bufferSize: 3390,
					buckets: 128,
					threads: 2,
					stripeSize: 65536,
					phase1: {startTime: new Date('2021-05-08T09:38:18.000Z')}
				});
				done();
			});

			parser.start();
		});

		it('emits error if file is not readable', (done) => {
			const parser = plot('.');
			parser.on('error', (err) => {
				expect(err.message).toEqual('EISDIR: illegal operation on a directory, read');
				done();
			});

			parser.start();
		});
	});
});


test/index.test.js
'use strict';

const expect = require('expect');
const logParser = require('../');
const logFile = __dirname + '/fixtures/plot-log.txt';
const logFilePart = __dirname + '/fixtures/plot-log-unfinished.txt';

describe('Plot parser', () => {

	describe('parse()', () => {

		it('parses log file', async () => {
			const parser = logParser(logFile);
			const res = await parser.parse();

			expect(res.modified).toBeInstanceOf(Date);
			expect(res.created).toBeInstanceOf(Date);
			delete res.created;
			delete res.modified;

			expect(res).toEqual({
				tmpDirs: ['/foobar', '/foobar'],
				id: '1553a1f4a9bb4d36a8a81e4334854c36d8abec1f192c362a28bd935fb035a03c',
				plotSize: 32,
				bufferSize: 3390,
				buckets: 128,
				threads: 2,
				stripeSize: 65536,
				phase1: {
					startTime: new Date('2021-05-08T09:38:18.000Z'),
					seconds: 9998.132,
					cpuPercent: 145.15,
					endTime: new Date('2021-05-08T12:24:56.000Z')
				},
				phase2: {
					startTime: new Date('2021-05-08T12:24:56.000Z'),
					seconds: 4188.688,
					cpuPercent: 98.23,
					endTime: new Date('2021-05-08T13:34:45.000Z')
				},
				phase3: {
					startTime: new Date('2021-05-08T13:34:45.000Z'),
					seconds: 7607.997,
					cpuPercent: 98.21,
					endTime: new Date('2021-05-08T15:41:33.000Z')
				},
				phase4: {
					startTime: new Date('2021-05-08T15:41:33.000Z'),
					seconds: 552.296,
					cpuPercent: 99.46,
					endTime: new Date('2021-05-08T15:50:45.000Z')
				},
				finalFileSize: 101.42,
				totalTimeSeconds: 22347.115,
				cpu: 119.24,
				finishedTime: new Date('2021-05-08T15:50:45.000Z'),
				copyTimeSeconds: 152.474,
				copyCpu: 75.12,
				copyFinishedTime: new Date('2021-05-08T15:53:19.000Z')
			});
		});

		it('parses non finished file', async () => {
			const parser = logParser(logFilePart);
			const res = await parser.parse();

			expect(res.modified).toBeInstanceOf(Date);
			expect(res.created).toBeInstanceOf(Date);
			delete res.created;
			delete res.modified;
			expect(res).toEqual({
				tmpDirs: ['/foobar', '/foobar'],
				id: '1553a1f4a9bb4d36a8a81e4334854c36d8abec1f192c362a28bd935fb035a03c',
				plotSize: 32,
				bufferSize: 3390,
				buckets: 128,
				threads: 2,
				stripeSize: 65536,
				phase1: {startTime: new Date('2021-05-08T09:38:18.000Z')}
			});
		});

	});

	describe('start()', () => {

		it('emits events', (done) => {
			const parser = logParser(logFile);
			const events = {
				phaseStart: 0,
				phaseEnd: 0,
				error: 0,
				started: 0,
				finished: 0,
				done: 0,
				progress: 0
			};

			Object.keys(events).forEach((event) => {
				parser.on(event, (...args) => {
					events[event]++;
					if (event === 'done') {
						expect(events).toEqual({
							phaseStart: 4,
							phaseEnd: 4,
							error: 0,
							started: 1,
							finished: 1,
							done: 1,
							progress: 19
						});
						done();
					}
				});
			});

			parser.start();
		});

		it('emits parseEnd for non finished log file', (done) => {
			const parser = logParser(logFilePart);
			parser.on('parseEnd', (data) => {
				delete data.created;
				delete data.modified;

				expect(data).toEqual({
					tmpDirs: ['/foobar', '/foobar'],
					id: '1553a1f4a9bb4d36a8a81e4334854c36d8abec1f192c362a28bd935fb035a03c',
					plotSize: 32,
					bufferSize: 3390,
					buckets: 128,
					threads: 2,
					stripeSize: 65536,
					phase1: {startTime: new Date('2021-05-08T09:38:18.000Z')}
				});
				done();
			});

			parser.start();
		});

		it('emits error if file is not readable', (done) => {
			const parser = logParser('.');
			parser.on('error', (err) => {
				expect(err.message).toEqual('EISDIR: illegal operation on a directory, read');
				done();
			});

			parser.start();
		});
	});
});


test/harvester.test.js
'use strict';

const expect = require('expect');
const harvester = require('../src/harvester');
const logFile = __dirname + '/fixtures/harvest-log.txt';

describe('Harvest parser', () => {

	describe('parse()', () => {

		it('parses log file', async () => {
			const parser = harvester(logFile);
			const res = await parser.parse();
			expect(res).toEqual({
				signagePoint: [{
					timestamp: new Date('2021-05-22T15:38:49.837'),
					eligible: 1,
					hash: '486436f706',
					proofs: 0,
					duration: 0.07757,
					plots: 66
				}, {
					timestamp: new Date('2021-05-22T15:39:06.358'),
					eligible: 0,
					hash: '486436f706',
					proofs: 0,
					duration: 0.00668,
					plots: 66
				}, {
					timestamp: new Date('2021-05-22T15:39:06.967'),
					eligible: 0,
					hash: '486436f706',
					proofs: 0,
					duration: 0.00504,
					plots: 66
				}],
				warning: [{
					timestamp: new Date('2021-05-21T09:11:48.229'),
					message: 'Directory: /Volumes/foobar/plots does not exist.'
				}, {
					timestamp: new Date('2021-05-22T11:00:22.259'),
					message: 'Looking up qualities on /Volumes/foobar/plots/plot-k32-2021-05-15-13-47-aff22113d6bbe9d1a37be3ee7ccb295d083085ff93a1942b1ccfbe7ffa41e5d4.plot took: 6.22216010093689. This should be below 5 seconds to minimize risk of losing rewards.'
				}],
				load: [{
					plots: 66,
					size: 6.532919835822213,
					sizeUnit: 'TiB',
					time: 0.08295607566833496,
					timestamp: new Date('2021-05-22T15:37:57.822')
				}]
			});
		});

	});

	describe('watch()', () => {

		it('emits events', (done) => {
			const parser = harvester(logFile);
			const events = {
				signagePoint: 0,
				warning: 0,
				load: 0
			};

			Object.keys(events).forEach((event) => {
				parser.on(event, (...args) => {
					events[event]++;
				});
			});

			parser.once('endParse', () => {
				parser.stop();
				expect(events).toEqual({
					warning: 2,
					signagePoint: 3,
					load: 1
				});
				done();
			});

			parser.watch(0).catch(done);
		});

	});
});


src/harvester.js
'use strict';

const fs = require('fs');
const EventEmitter = require('events');
const TailFile = require('@logdna/tail-file');
const split2 = require('split2');
const {mapFilterValues} = require('./lib/filter');

const filters = [{
	regexp: /^([^\s]+) harvester chia.harvester.harvester: INFO\s+(\d+) plots were eligible for farming ([^\.]+)\.\.\. Found (\d+) proofs. Time: (\d+\.\d+) s. Total (\d+) plots/,
	values: [['timestamp', (d) => new Date(d)], ['eligible', Number], 'hash', ['proofs', Number], ['duration', Number], ['plots', Number]],
	event: 'signagePoint'
}, {
	regexp: /^([^\s]+) harvester [^:]+: WARNING\s+([^$]+)/,
	values: [['timestamp', (d) => new Date(d)], 'message'],
	event: 'warning'
}, {
	regexp: /^([^\s]+) harvester [^:]+: INFO\s+Loaded a total of (\d+) plots of size (\d+(\.\d+)?) ([^,]+), in (\d+(\.\d+)?)/,
	values: [['timestamp', (d) => new Date(d)], ['plots', Number], ['size', Number],, 'sizeUnit', ['time', Number]],
	event: 'load'
}];

class HarvestParser extends EventEmitter {
	constructor(file) {
		super();
		this.file = file;
		this.isWatching = false;
	}

	async watch(startPos) {
		if (this.isWatching) {
			throw new Error('Watch is already running');
		}

		this.tail = new TailFile(this.file, {startPos});
		this.tail.on('tail_error', (err) => {
			this.emit('error', err);
		});

		this.tail.on('error', (err) => {
			this.emit('error', err);
		});

		this.tail.on('flush', () => {
			this.emit('endParse');
		});

		await this.tail.start();
		this.isWatching = true;
		this.tail
			.pipe(split2())
			.on('data', (line) => this.onLine(line));
	}

	stop() {
		this.tail.quit();
		this.isWatching = false;
	}

	parse() {
		let resolve, reject, error;
		const promise = new Promise((res, rej) => {
			resolve = res;
			reject = rej;
		});

		const result = {
			signagePoint: [],
			warning: [],
			load: []
		};

		const onError = (err) => {
			error = err;
		};

		const update = (event) => (data) => result[event].push(data);
		const updateEligable = update('signagePoint');
		const updateWarning = update('warning');
		const updateLoad = update('load');

		const done = () => {
			this.off('signagePoint', updateEligable);
			this.off('warning', updateWarning);
			this.off('load', updateLoad);
			this.off('error', onError);
			if (error) {
				return reject(error);
			}

			resolve(result);
		};

		this.on('signagePoint', updateEligable);
		this.on('warning', updateWarning);
		this.on('load', updateLoad);
		this.on('error', onError);
		fs.createReadStream(this.file)
			.on('error', (err) => {
				error = err;
			})
			.pipe(split2())
			.on('error', (err) => {
				error = err;
			})
			.on('data', (line) => this.onLine(line))
			.on('end', done);

		return promise;
	}

	onLine(line) {
		let matches;
		const filter = filters.find((f) => {
			matches = f.regexp.exec(line);
			return matches;
		});

		if (!matches) {
			return;
		}

		const data = mapFilterValues(filter.values, matches, {});
		if (filter.event) {
			this.emit(filter.event, data);
		}
	}
}

module.exports = (file) => {
	return new HarvestParser(file);
};

module.exports.HarvestParser = HarvestParser;


src/fileparser.js
'use strict';
const fs = require('fs');
const EventEmitter = require('events');
const split2 = require('split2');
const TailFile = require('@logdna/tail-file');
const {mapFilterValues} = require('./lib/filter');

class FileParser extends EventEmitter {
	constructor(file) {
		super();
		this.file = file;
		this.isWatching = false;
	}

	parse() {
		return new Promise((resolve, reject) => {
			function error(err) {
				this.off('parseEnd', done);
				reject(err);
			}

			function done(data) {
				this.off('error', error);
				resolve(data);
			}

			this.once('error', error);
			this.once('parseEnd', done);
			this.start();
		});
	}

	async watch() {
		if (this.isWatching) {
			throw new Error('Watch is already running');
		}

		this.reset();
		try {
			await this.onStart();
		} catch (err) {
			this.emit('error', err);
			return;
		}

		this.tail = new TailFile(this.file, {startPos: 0});
		this.tail.on('tail_error', (err) => {
			this.emit('error', err);
		});

		this.tail.on('error', (err) => {
			this.emit('error', err);
		});

		await this.tail.start();
		this.isWatching = true;
		this.tail.pipe(split2()).on('data', (line) => {
			this.emit('line', line);
			if (this.isFinished()) {
				this.tail.quit();
				this.emit('done', this.data);
			}
		});
	}

	async start() {
		this.reset();

		try {
			await this.onStart();
		} catch (err) {
			this.emit('error', err);
			return;
		}

		fs.createReadStream(this.file)
			.on('error', (err) => this.emit('error', err))
			.pipe(split2())
			.on('error', (err) => this.emit('error', err))
			.on('data', (line) => {
				this.emit('line', line);
			})
			.on('end', () => {
				this.emit('parseEnd', this.data);
				if (this.isFinished()) {
					this.emit('done', this.data);
				}
			});
	}

	reset() {
		this.data = {};
	}

	onStart() {
		throw new Error('not implemented');
	}

	isFinished() {
		throw new Error('not implemented');
	}

	onFilter(filter, match, data) {
		mapFilterValues(filter.values, match, data);

		if (filter.event) {
			this.emit(filter.event, data);
		}
	}

}

module.exports = FileParser;


src/plot.js
'use strict';
const fs = require('fs');
const {stat: fsStat} = fs.promises;
const {parse: parseDate} = require('date-fns');
const FileParser = require('./fileparser');

const startPlot = [
	{
		regexp: /Starting plotting progress into temporary dirs: ([^\s]+) and ([^\s]+)/,
		values: [['tmpDirs'], ['tmpDirs']]
	},
	{
		regexp: /ID: ([^\s]+)/,
		values: ['id']
	},
	{
		regexp: /Plot size is: (\d+)/,
		values: [['plotSize', Number]]
	},
	{
		regexp: /Buffer size is: (\d+)/,
		values: [['bufferSize', Number]]
	},
	{
		regexp: /Using (\d+) buckets/,
		values: [['buckets', Number]]
	},
	{
		event: 'started',
		regexp: /Using (\d+) threads of stripe size (\d+)/,
		values: [['threads', Number], ['stripeSize', Number]]
	}
];

const endPlot = [
	{
		regexp: /Final File size: (\d+(\.\d+)?)/,
		values: [['finalFileSize', Number]]
	},
	{
		regexp: /Total time = (\d+(\.\d+)?) seconds. CPU \((\d+(\.\d+)?)%\) ([^$]+)/,
		values: [['totalTimeSeconds', Number],, ['cpu', Number],, ['finishedTime', stringToDate]]
	},
	{
		event: 'finished',
		regexp: /Copy time = (\d+(\.\d+)?) seconds. CPU \((\d+(\.\d+)?)%\) ([^$]+)/,
		values: [['copyTimeSeconds', Number],, ['copyCpu', Number],, ['copyFinishedTime', stringToDate]]
	}
];

const startPhase = /Starting phase (\d+)\/(\d+): .*?\.\.\.\s([^$]+)/;
const endPhase = /Time for phase (\d+) = (\d+(\.\d+)?) seconds. CPU \((\d+(\.\d+)?)%\) ([^$]+)/;

const progess = [
	[/Computing table (\d+)/, [1, 6, 12, 20, 28, 36, 42]],
	[/Backpropagating on table (\d+)/, [63, 61, 58, 55, 51, 48, 43]],
	[/Compressing tables (\d+)/, [66, 73, 79, 85, 92, 98]]
];

function stringToDate(str) {
	return parseDate(str.replace(/\s{2}/g, ' '), 'EEE MMM d HH:mm:ss yyyy', new Date());
}

class PlotParser extends FileParser {
	constructor(file) {
		super(file);
		this.on('line', (line) => this.onNewLine(line));
	}

	async onStart() {
		const fileStat = await fsStat(this.file);
		this.data.created = fileStat.birthtime;
		this.data.modified = fileStat.mtime;
	}

	reset() {
		super.reset();
		this.startIndex = 0;
		this.endIndex = 0;
		this.currentPhase = 0;
		this.started = false;
		this.finished = false;
	}

	onNewLine(line) {
		if (!line.length || this.finished) {
			return;
		}

		if (!this.started) {
			const stage = startPlot[this.startIndex];
			const match = stage.regexp.exec(line);
			if (match) {
				this.startIndex++;
				this.onFilter(stage, [...match], this.data);
				this.started = stage.event === 'started';
			}
			return;
		}

		this.checkForPhase(line);

		if (this.currentPhase && progess[this.currentPhase - 1]) {
			this.checkPhaseProgress(line, progess[this.currentPhase - 1]);
		}

		if (this.data.phase4?.endTime) {
			const stage = endPlot[this.endIndex];
			const match = stage.regexp.exec(line);
			if (match) {
				this.endIndex++;
				this.onFilter(stage, [...match], this.data);
				this.finished = stage.event === 'finished';
			}
		}
	}

	isFinished() {
		return this.finished;
	}

	checkForPhase(line) {
		const start = startPhase.exec(line);
		if (start) {
			const phase = this.currentPhase = start[1];
			const data = {
				startTime: stringToDate(start[3])
			};

			this.data[`phase${this.currentPhase}`] = data;
			this.emit('phaseStart', phase, data);
			return;
		}

		const end = endPhase.exec(line);
		if (end) {
			const phase = end[1];
			const data = {
				seconds: Number(end[2]),
				cpuPercent: Number(end[4]),
				endTime: stringToDate(end[6])
			};

			Object.assign(this.data[`phase${end[1]}`], data);
			this.emit('phaseEnd', phase, data);
			return;
		}
	}

	checkPhaseProgress(line, [regexp, percents]) {
		const match = regexp.exec(line);
		if (match) {
			this.emit('progress', percents[match[1] - 1]);
		}
	}
}

module.exports = function (file) {
	return new PlotParser(file);
};

module.exports.PlotParser;
if (process.env.NODE_ENV === 'test') {
	module.exports.stringToDate = stringToDate;
}
// const parser = new PlotParser('/Users/martin/.chia/mainnet/plotter/pny_2021-05-21_13_38_03_581568.log');
// ['phaseStart', 'phaseEnd', 'error', 'started', 'finished', 'done', 'progress'].forEach((event) => {
// 	parser.on(event, (...args) => console.log(`[${event}]`, ...args));
// });

// parser.watch().catch(console.log);
